tests/evm/test_error_oog_memory_copy.py

```diff
@@ -0,0 +1,153 @@
+    # constant gas cost of the following opcodes is 3
+    (False, False, Opcode.CALLDATACOPY, 32, 32, 11),
+    (False, False, Opcode.CODECOPY, 32, 32, 11),
+    (False, False, Opcode.RETURNDATACOPY, 32, 32, 11),
+    (True, False, Opcode.CALLDATACOPY, 32, 32, 11),
+    (True, False, Opcode.CODECOPY, 32, 32, 11),
+    (True, False, Opcode.RETURNDATACOPY, 32, 32, 11),
+    # EXTCODECOPY
+    # constant gas cost of EXTCODECOPY is either 100 (warm access) or 2600 (cold access)
+    (False, False, Opcode.EXTCODECOPY, 32, 32, 2608),
+    (False, True, Opcode.EXTCODECOPY, 32, 32, 108),
+    (True, False, Opcode.EXTCODECOPY, 32, 32, 2608),
+    (True, True, Opcode.EXTCODECOPY, 32, 32, 108),
+)
+
+
+@pytest.mark.parametrize(
+    "is_root, is_warm_access, opcode, offset, length, gas_left", TESTING_DATA_IS_ROOT
+)
+def test_error_oog_memory_copy(
+    is_root: bool, is_warm_access: bool, opcode: Opcode, offset: int, length: int, gas_left: int
+):
+    caller_id = 1 if is_root else 2
+    is_ext_code_copy = True if opcode == Opcode.EXTCODECOPY else False
+
+    # warm/cold access only occurred while EXTCODECOPY
+    if is_warm_access:
+        assert is_ext_code_copy
+
+    if is_ext_code_copy:
+        rw_counter = 9 if is_root else 20
+        address = 0xCAFECAFE
+        bytecode = Bytecode().extcodecopy()
+        rw_table = (
+            RWDictionary(rw_counter)
+            .stack_read(caller_id, 1020, Word(address))
+            .stack_read(caller_id, 1021, Word(offset))
+            .stack_read(caller_id, 1022, Word(0))
+            .stack_read(caller_id, 1023, Word(length))
+        )
+        # fmt: off
+        rw_table \
+            .call_context_read(caller_id, CallContextFieldTag.TxId, caller_id) \
+            .call_context_read(caller_id, CallContextFieldTag.RwCounterEndOfReversion, 1) \
+            .call_context_read(caller_id, CallContextFieldTag.IsPersistent, False) \
+            .tx_access_list_account_write(caller_id, address, True, is_warm_access, rw_counter_of_reversion=1)
+        # fmt: on
+        stack_pointer = 1020
+        pc = 133
+    else:
+        rw_counter = 3
+        if opcode == Opcode.CALLDATACOPY:
+            bytecode = Bytecode().calldatacopy()
+        elif opcode == Opcode.CODECOPY:
+            bytecode = Bytecode().codecopy()
+        else:
+            bytecode = Bytecode().returndatacopy()
+
+        rw_table = (
+            RWDictionary(rw_counter)
+            .stack_read(caller_id, 1021, Word(offset))
+            .stack_read(caller_id, 1022, Word(0))
+            .stack_read(caller_id, 1023, Word(length))
+        )
+        stack_pointer = 1021
+        pc = 100
+
+    bytecode_hash = Word(bytecode.hash())
+
+    rw_table.call_context_read(caller_id, CallContextFieldTag.IsSuccess, 0)
+
+    # fmt: off
+    if not is_root:
+        memory_word_size = (
+            0 if length == 0 else (offset + length + 31) // 32
+        )
+        rw_table \
```

Comment:
Is that an extra row for STOP? If so, none are needed for EndTx?

Reply:
From my understanding, if an error happens, we need to restore the context back to the context of the parent call. That's why we need the following lookups for checking the call context. I'm not familiar BeginTx and EndTx enough, but there is no parent call for root calls. I'm assumming that's why we don't need to check these things.